eox_tenant/auth.py

```diff
@@ -0,0 +1,95 @@
+
+"""
+This file implements the authentication backend for the openedx platform.
+"""
+import logging
+import re
+
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.contrib.auth.backends import AllowAllUsersModelBackend as UserModelBackend
+
+from django.utils.translation import ugettext as _
+
+from ratelimitbackend.backends import RateLimitMixin
+
+from eox_tenant.edxapp_wrapper.auth import get_edx_auth_failed
+
+AuthFailedError = get_edx_auth_failed()
+AUDIT_LOG = logging.getLogger("audit")
+UserModel = get_user_model()
+
+
+class TenantAwareAuthBackendBase(UserModelBackend):
+    """
+    Authentication Backend class which will check if the user has a signupsource in the requested site.
+    """
+    def authenticate(self, request, username=None, password=None, **kwargs):
```

Comment:
@morenol I think we can use the super method here to avoid replicating the code from django. What we can do is overwriting the user_can_authenticate method and there we can call the user_can_authenticate_on_tenant method to be a bit more compliant with the django backend structure

Reply:
@morenol in fact the only thing we need is to ovwerwrite the method `user_can_authenticate`

Reply:
That was my first approach, but the user_can_authenticate method does not receive the request which is being used to get the HTTP_HOST

https://github.com/django/django/blob/bc1f79d0a01d085500aa82ff29800403291a91a4/django/contrib/auth/backends.py#L109

Reply:
@morenol one option could be overwriting `authenticate` method and in the first line make something like `self.request = request`, from that point it's available on any method, but I don't know what do you think about it

Reply:
That is something that we can do, but I don't know if that is better than the current change

Reply:
The benefits of that is that we no longer will depend on ratelimitbackend

Reply:
Here I see two benefits: 1) not replicating the code from django, 2) avoid the dependency on ratelimitbackend. My approach here would be:

- Inherit directly from the edx-platform backend (I already tested the `super` calls and it should work as expected)
- Overwrite the `authenticate` method, add the line `self.request = request` and return the result of the `super` call
- Overwrite the `user_can_authenticate` method and inside it, call the `user_can_authenticate_on_tenant` method based on the feature flag EDNX_TENANT_AWARE_AUTH.

Reply:
I made those changes (A little bit different the last item)